### ISSUE TITLE
Google Batch: Input files provided as output to the next process are null (mounted symlink not translated into real file on output)

### ISSUE THREAD

**Issue by VasLem** [nextflow-io/nextflow#4819]
### Discussed in https://github.com/nextflow-io/nextflow/discussions/4814

<div type='discussions-op-text'>

<sup>Originally posted by **VasLem** March 13, 2024</sup>
I have the following process:

```
process readFromDirectory {
    input:
        path sample_dir
        val sample_name
    output:
        tuple val(sample_name), path("output/*.*")

    script:
        """
        mkdir -p output
        export FILES=`find $sample_dir/ -print0 | perl -0ne "print if /${sample_name}.*\\.f(?:ast)?[qa](?:\\.gz)?/s"`
        ln -sf \$FILES output/
        """    
}
```

which is supposed to read a folder, and return the files that match to the output. This works locally but not in the google cloud, as the symbolic link ends up dangling. I feel this is a bug, or at least there is lack of documentation explaining this edge case. `followLinks`  is by default true, therefore one would expect to get the files and not the links to the next process, particularly not dangling ones... I have tried also other ways such as this:

```
process readFromDirectory {
    input:
        path sample_dir
        val sample_name
    output:
        tuple val(sample_name), path(files)

    script:
        files = sample_dir.find { it.toString() =~  /${sample_name}.*\.f(?:ast)?[qa](?:\.gz)?/ }
        """
        """    
}
```

which fails as the files is null (!!!!)

or even by transforming the findings into actual file names (prepending the gs bucket directory and then trying to convert that to links), like here:

```
process readFromDirectory {
    input:
        path sample_dir
        val sample_dir2
        val sample_name

    script:
        """
        parent_dir=`dirname $sample_dir2`
        
        export FILES=`find $sample_dir/ -print0 | perl -0ne "print if /${sample_name}.*\\.f(?:ast)?[qa](?:\\.gz)?/s" | xargs -r0 -I {} echo \$parent_dir/{}`
        """

    output:
        tuple val(sample_name), env(FILES)     
}

workflow {
        rf = readFiles(sample_dir, sample_dir, samples).collect().toList()
        Channel.fromFilePairs("$sample_dir/*", size: -1) { file -> { if file in} }
        samples_and_reads = rf.map{
            it -> 
                [it[0], it[1].split(" ").each{v -> file(v)}]
            }
}
```
which again fails, returning that the string (which is not a string but a list inside the list) cannot be considered as path.
</div>

**Comment by harper357** [comment]
I'll add it here too, I actually ended up using `stageInMode = 'copy'` to fix things. Retesting things this morning, `stageInMode='link'` didn't end up working, it gave a different error.

**Comment by VasLem** [comment]
I had a similar experience with @harper357 . To be able to use Nextflow for such cases, where the input propagates to the output, I had to dereference the input, like the following example, where I publish an incoming set of files to a folder:
```
process publishToDir {

    disk { us = 10.GB + 2.B * reads*.size().sum(); println "Disk used by publishToDir: $us"; us }

    input:
        tuple val(sample_name), path(reads, arity:'1..*')
        val pub_dir
    
    publishDir "$pub_dir", mode: 'copy'

    output:
        path "*.*", includeInputs: true
    
    script:
        """
          for read in $reads
          do
                  if [ -L \${read} ]
                  then
                      echo "Replacing \$read symlink to actual file.."
                      cp --remove-destination `readlink \$read` \$read
                  else
                      "\$read is an actual file with size: \$(du -bs \$read | awk '{\$1=\$1/2^30"GB"}1')"
                  fi
          done
        echo "Publishing $reads to $pub_dir"
        """
}
```

I am going to try the `stageInMode` option when I have some time, however I am guessing if I set it to `'copy'`, it will just do the exact same thing that I am doing, under the hood.

**Comment by archmageirvine** [comment]
Also encountered this bug. Setting `stageInMode` to `copy` in the first process worked as a fix for me, but the solution is not ideal given potential for large files.

**Comment by AugustDev** [comment]
Problem still persists, any solutions?

**Comment by VasLem** [comment]
I have set all the flags (`stageInMode` `stageOutMode` and `mode`) to 'copy' , and whenever I have `includeInputs: true` in the output path, I am getting 0 bytes files in the directory I am publishing, if those files have been staged for that process. It has been long that nothing has been done for this, so I guess people have managed to circumvent it somehow? It must be affecting many Google Batch users.

**Comment by harper357** [comment]
@VasLem  are you setting those flags on all processes, the process that is crashing, or the process before the process crashing?

I am no longer at the job where I was using that pipeline, so I can't double check, but I believe that I made my change on all processes. In other words, in `nextflow.config`, `conf/base.config`, or your custom `conf/personal.config` :

```
process {
    stageInMode = "copy"
}
```

**Comment by VasLem** [comment]
Hi @harper357 , I am only doing it exclusively for one process, which needs to be able take in existing files so that not to redo the entire computation. At a local context, this runs flawlessly. However, when I add `includeInputs: true` on google batch, the already processed files are replaced by 0 bytes ones. I would add the `optional` flag to the output and remove `includeInputs` but then I would need to include a new file to report the success of the process. All the possibilities appear so ugly :(.

**Comment by harper357** [comment]
Maybe I am misunderstanding, but you have Task 1 that generates output A and then Task 2 and Task 3 that both need output A? That sounds like you should be using a channel to pass output A directly from Task 1  to both Task 2 and Task 3. 

So it would look something like : 
```
TASK_1(inputs)
TASK_2(TASK_1.out.A, foo_inputs)
TASK_3(TASK_1.out.A, bar_inputs)
```

**Comment by VasLem** [comment]
Not exactly, it is the exact same process that allows partial processing of the input up to a point, for debugging purposes, something like:
```
out_dir = "out/*"
exfiles = file(out_dir).listFiles().findAll{x-> x.size() > 0}.collect{x-> x.toUriString()}}
output = task(inputs, exfiles, "until_step_a")
exfiles = file(out_dir).listFiles().findAll{x-> x.size() > 0}.collect{x-> x.toUriString()}}
output = task(inputs, exfiles , "until_step_b")
```
The reason is that if the step b fails, I never get to see step a results, as the entire google VM is thrown away (another bad thing about Nextflow and Google Batch, but I think this is an issue from Google's side).

**Comment by VasLem** [comment]
The issue has not been solved. Relevant discussion: https://github.com/nextflow-io/nextflow/discussions/6185
